### PR TITLE
feature: Implement Enhanced Retry Logic for API calls

### PR DESCRIPTION
**Pull request by borisprimer**
[JIRA](https://primerapi.atlassian.net/browse/ACC-3652)
- implemented retry feature with a configuration object (retry can be applied to any API call)
- added logs
- added analytics
- added some tests

QA NOTES: To reproduce retry I intentionally changed the base url of the /configuration endpoint by adding "test" to it. Since that will return 404, I changed my code related to 500error in code to check against 400,  `.statusCode >= 500` -> `.statusCode >= 400`.

### CODE REVIEW

File: Sources/PrimerSDK/Classes/PCI/Services/RetryHandler.swift [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+            let metadata = ResponseMetadataModel(responseUrl: httpResponse.url?.absoluteString,
+                                                 statusCode: httpResponse.statusCode,
+                                                 headers: httpResponse.allHeaderFields as? [String: String])
+            let responseModel = DispatcherResponseModel(metadata: metadata, requestDuration: requestDuration, data: data, error: error)
+
+            // Check if the response is successful
+            if (200...299).contains(responseModel.metadata.statusCode) {
+                let successMessage = "Request succeeded after \(self.retries) retries. Status code: \(responseModel.metadata.statusCode)"
+                self.logger.debug(message: successMessage)
+                let retryEvent = Analytics.Event.message(message: successMessage, messageType: .info, severity: .info)
```
Comment (NQuinn27) [inline Sources/PrimerSDK/Classes/PCI/Services/RetryHandler.swift:97]: EventType should be `RETRY_SUCCESS` based on [this RFC](https://www.notion.so/primerio/Analytics-events-Retries-and-caching-f9935591702e4d1ead70c3b7891a43c6?pvs=4#d82d1f64eef44230b0ae06956b318b27).

All logs/events should be checked for alignment

Comment (borisprimer) [reply Sources/PrimerSDK/Classes/PCI/Services/RetryHandler.swift]: Done!

File: Sources/PrimerSDK/Classes/Extensions & Utilities/Error+Extensions.swift [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+//  Error+Extensions.swift
+//  PrimerSDK
+//
+//  Created by Boris on 25.7.24..
+//
+
+import Foundation
+
+internal extension Error {
+    var isNetworkError: Bool {
```
[the remark this answers is not in the document]
Comment (borisprimer) [reply Sources/PrimerSDK/Classes/Extensions & Utilities/Error+Extensions.swift:11]: Are you saying I should add more code checks to justify the name?

File: Sources/PrimerSDK/Classes/PCI/Services/RetryHandler.swift [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        self.request = request
+        self.retryConfig = retryConfig
+        self.completion = completion
+        self.urlSession = urlSession
+    }
+
+    func calculateBackoffWithJitter(baseDelay: TimeInterval, retryCount: Int, maxJitter: TimeInterval) -> TimeInterval {
+        let exponentialPart = baseDelay * pow(2.0, Double(retryCount - 1))
+        let jitterPart = Double.random(in: 0...maxJitter)
+        return min(exponentialPart + jitterPart, Double.greatestFiniteMagnitude)
```
[the remark this answers is not in the document]
Comment (borisprimer) [reply Sources/PrimerSDK/Classes/PCI/Services/RetryHandler.swift:29]: 100% Right. Thanks!

File: Sources/PrimerSDK/Classes/PCI/Services/DefaultNetworkService.swift [hunk 4]
```diff
@@ -57,8 +55,48 @@ class DefaultNetworkService: NetworkService, LogReporter {
...
+    }
+
+    @discardableResult
+    func request<T: Decodable>(_ endpoint: Endpoint,
+                               completion: @escaping ResponseCompletionWithHeaders<T>) -> PrimerCancellable? {
+        return request(endpoint, retryConfig: nil, completion: completion)
+    }
+
+    @discardableResult
+    func request<T: Decodable>(_ endpoint: Endpoint,
```
[the remark this answers is not in the document]
Comment (borisprimer) [reply Sources/PrimerSDK/Classes/PCI/Services/DefaultNetworkService.swift]: removed.

File: Sources/PrimerSDK/Classes/Extensions & Utilities/Error+Extensions.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (borisprimer) [reply Sources/PrimerSDK/Classes/Extensions & Utilities/Error+Extensions.swift:11]: I added other codes, please review again.

File: Sources/PrimerSDK/Classes/PCI/Services/RetryHandler.swift [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        if self.retries >= self.retryConfig.maxRetries {
+            errorMessage += "Reached maximum retries (\(self.retryConfig.maxRetries)).\nLast error object: \(error?.localizedDescription ?? "nil")"
+        } else if responseModel.metadata.statusCode >= 500 {
+            errorMessage += "Server error: \(responseModel.metadata.statusCode)"
+        } else if let error = error {
+            errorMessage += "Last error object: \(error.localizedDescription)"
+        } else {
+            errorMessage += "Status code: \(responseModel.metadata.statusCode)"
+        }
+        self.logger.debug(message: errorMessage)
```
Comment (semirp) [inline Sources/PrimerSDK/Classes/PCI/Services/RetryHandler.swift:66]: let's report `error` logs here

File: Sources/PrimerSDK/Classes/PCI/Services/RetryHandler.swift [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+            } else if is500Error {
+                finalErrorMessage += "HTTP \(response.metadata.statusCode) error encountered and RetryConfig.retry500Errors is \(retryConfig.retry500Errors ? "enabled" : "disabled")."
+            } else if isSomeOtherError {
+                finalErrorMessage += "HTTP \(response.metadata.statusCode) error encountered."
+            }
+
+            finalErrorMessage += "Last error: \(error?.localizedDescription ?? "Unknown error")"
+            let retryEvent = Analytics.Event.message(message: finalErrorMessage, messageType: .retry, severity: .warning)
+            Analytics.Service.record(event: retryEvent)
+            self.logger.debug(message: finalErrorMessage)
```
Comment (semirp) [inline Sources/PrimerSDK/Classes/PCI/Services/RetryHandler.swift:129]: let's report warning log here

**Review by semirp — APPROVED** [review]

**Pull request merged**